internal/controller/object/object.go

```diff
@@ -340,3 +354,160 @@ func setObserved(obj *v1alpha1.Object, observed *unstructured.Unstructured) erro
 	}
 	return nil
 }
+
+// resolveReferencies resolves references for the current Object. If it fails to
+// resolve some reference, e.g.: due to reference not ready, it will then return
+// error and requeue to wait for resolving it next time.
+func (c *external) resolveReferencies(ctx context.Context, obj *v1alpha1.Object) error {
+	c.logger.Debug("Resolving referencies.")
+
+	// Loop through references to resolve each referenced resource
+	for _, ref := range obj.Spec.References {
+		var kubeClient client.Client
+
+		if ref.FromObject.APIVersion == obj.APIVersion && ref.FromObject.Kind == obj.Kind {
+			// Object resource handled by provider on local cluster
+			kubeClient = c.localClient
+		} else {
+			// Other resource usually handled on remote cluster
+			// But could also be handled on local cluster
+			kubeClient = c.client
+		}
```

Comment:
IIUC, this seems a bit problematic to me. 
- How could we decide to use localclient by just checking if `APIVersion` and `Kind` matches for object itself and it references? 

My understanding so far is, we will always use localclient for dependencies and references. If we need to reference a remote resource, e.g. a K8S CR, we will need to create a local Object CR tracking it with Observe/ObserveDelete policy and resolve references from it.

Am I missing something, could you elaborate a bit?

Reply:
You are right! My original intention was to support both Object and arbitrary K8s CR. For remote CR, it will be captured by Object from local. So, yes, all should happen at local. I've updated the code to reflect this. It actually simplifies the code.